src/python/pants/backend/python/rules/pytest_coverage.py

```diff
@@ -246,98 +260,74 @@ async def setup_coverage(coverage: PytestCoverage) -> CoverageSetup:
     return CoverageSetup(requirements_pex)
 
 
-@dataclass(frozen=True)
-class PytestCoverageDataBatch(CoverageDataBatch):
-    addresses_and_test_results: Tuple[AddressAndTestResult, ...]
-
-
 @dataclass(frozen=True)
 class MergedCoverageData:
     coverage_data: Digest
 
 
-@named_rule(desc="Merge coverage reports")
+@named_rule(desc="Merge Pytest coverage reports")
 async def merge_coverage_data(
-    data_batch: PytestCoverageDataBatch,
+    data_collection: PytestCoverageDataCollection,
+    coverage_setup: CoverageSetup,
     transitive_targets: TransitiveHydratedTargets,
     python_setup: PythonSetup,
-    coverage_setup: CoverageSetup,
     subprocess_encoding_environment: SubprocessEncodingEnvironment,
 ) -> MergedCoverageData:
-    """Takes all python test results and merges their coverage data into a single sql file."""
-    # We start with a bunch of test results, each of which has a coverage data file called `.coverage`
-    # We prefix each of these with their address so that we can write them all into a single pex.
+    """Takes all Python test results and merges their coverage data into a single SQL file."""
+    # We start with a bunch of test results, each of which has a coverage data file called
+    # `.coverage`. We prefix each of these with their address so that we can write them all into a
+    # single PEX.
     coverage_directory_digests = await MultiGet(
-        Get[Digest](
-            DirectoryWithPrefixToAdd(
-                directory_digest=result.test_result.coverage_data.digest,  # type: ignore[attr-defined]
-                prefix=result.address.path_safe_spec,
-            )
-        )
-        for result in data_batch.addresses_and_test_results
-        if result.test_result.coverage_data is not None
+        Get[Digest](DirectoryWithPrefixToAdd(data.digest, prefix=data.address.path_safe_spec))
+        for data in data_collection
```

Comment:
Simplified thanks to changing the definitions of `PytestCoverageData` and `PytestCoverageDataCollection` to only store the information they need.